lib/icingadb/icingadb-worker.cpp

```diff
@@ -0,0 +1,503 @@
+{
+}
+
+RelationsDeletionItem::RelationsDeletionItem(const String& id, RelationsKeyMap relations)
+	: PendingQueueItem{id, 0}, Relations{std::move(relations)}
+{
+}
+
+/**
+ * Background worker thread procedure for processing pending items.
+ *
+ * This function runs in a separate thread and continuously processes pending items that have been
+ * enqueued for Redis updates. It waits for new items to be added to the pending items container,
+ * and processes them one at a time, ensuring that the Redis connection is active and not overloaded
+ * with too many pending queries. The function also implements a delay mechanism to allow for potential
+ * additional changes to be merged into the same item before processing it.
+ */
+void IcingaDB::PendingItemsThreadProc()
+{
+	using namespace std::chrono_literals;
+	namespace ch = std::chrono;
+
+	// Limits the number of pending queries the Rcon can have at any given time to reduce the memory overhead to
+	// the absolute minimum necessary, since the size of the pending queue items is much smaller than the size
+	// of the actual Redis queries. Thus, this will slow down the worker thread a bit from generating too many
+	// Redis queries when the Redis connection is saturated.
+	constexpr size_t maxPendingQueries = 128;
+
+	std::unique_lock lock(m_PendingItemsMutex);
+	// Wait until the initial config dump is done. IcingaDB::OnConnectedHandler will notify us once it's finished.
+	while (GetActive() && !m_ConfigDumpDone) m_PendingItemsCV.wait(lock);
+
+	// Predicate to determine whether the worker thread is allowed to process pending items.
+	auto canContinue = [this] {
+		if (!GetActive()) {
+			return true;
+		}
+		return !m_PendingItems.empty() && m_RconWorker && m_RconWorker->IsConnected() && m_RconWorker->GetPendingQueryCount() < maxPendingQueries;
+	};
+
+	while (GetActive()) {
+		// Even if someone notifies us, we still need to verify whether the precondition is actually fulfilled.
+		// However, in case we don't receive any notification, we still want to wake up periodically on our own
+		// to check whether we can proceed (e.g. the Redis connection might have become available again and there
+		// was no activity on the pending items queue to trigger a notification). Thus, we use a timed wait here.
+		while (!canContinue()) m_PendingItemsCV.wait_for(lock, 100ms);
+
+		// If we're not active anymore (shutting down) and the Redis connection is gone, we can't process anything
+		// anymore, so just exit right away. If we still have a working connection, we want to flush all remaining
+		// items as fast as possible before exiting.
+		if (!GetActive() && (!m_RconWorker || !m_RconWorker->IsConnected())) {
+			break;
+		}
+
+		ch::duration<double> retryAfter;
+		do {
+			retryAfter = DequeueAndProcessOne(lock);
+			// Flush all remaining items as fast as possible during shutdown.
+			// At this point, there shouldn't be any new items being enqueued anymore,
+			// so we won't end up in an infinite loop.
```

Comment:
What would happen if we just abandon this queue on shutdown? Is there anything of value in there? Wouldn't the next full sync overwrite everything we would potentially write here?